### PR TITLE
Allow ATLAddressBarViewController delegates to specify prefill participants

### PR DESCRIPTION
**Pull request by dfinzer**
This supports https://github.com/layerhq/Atlas-iOS/issues/968

@kmahal let me know if there's anything I need to change here.

### CODE REVIEW

**Comment by dfinzer** [comment]
@kmahal @blakewatters wanted to check in and see if this change sounds reasonable

File: Code/Controllers/ATLAddressBarViewController.m [hunk 1]
```diff
@@ -384,6 +382,26 @@ - (void)searchEnded
...
     self.tableView.hidden = YES;
     [self.tableView reloadData];
+    [self prefillParticipants];
+}
+
+- (void)prefillParticipants
+{
+    if ([self.delegate respondsToSelector:@selector(addressBarViewController:retreivePrefilledParticipantsWithCompletion:)]) {
+        [self.delegate addressBarViewController:self retreivePrefilledParticipantsWithCompletion:^(NSArray *participants) {
+            if (!self.addressBarView.addressBarTextView.text.length) {
```
Comment (kmahal) [inline Code/Controllers/ATLAddressBarViewController.m]: With this line, after you select 1 participant to insert into the address bar, the table view will never be preloaded.  Is that the expected behavior?  After each insertion the user would expect to see the default list again.

File: Code/Controllers/ATLAddressBarViewController.h [hunk 2]
```diff
@@ -74,14 +74,21 @@
...
  @param searchText The text upon which a participant search should be performed.
  @param completion The completion block to be called upon search completion.
  @discussion Search should be performed across each `ATLParticipant` object's fullName property.
  */
 - (void)addressBarViewController:(ATLAddressBarViewController *)addressBarViewController searchForParticipantsMatchingText:(NSString *)searchText completion:(void (^)(NSArray *participants))completion;
 
+/**
+ @abstract Asks the data source for an NSArray of participants that will be used to prefill the participant list.
+ @param addressBarViewController The `ATLAddressBarViewController` to prefill.
+ @param completion The completion block to be called upon retreival of the participants.
```
Comment (kmahal) [inline Code/Controllers/ATLAddressBarViewController.h]: typo, spell it `retrieval`.

Comment (kmahal) [inline Code/Controllers/ATLAddressBarViewController.h]: add a @discussion that notes when this should or would be used.

**Comment by kmahal** [comment]
HI @dfinzer I've added a few comments for you to look at.  Thanks for the PR!

File: Code/Controllers/ATLAddressBarViewController.h [hunk 3]
```diff
@@ -74,14 +74,21 @@
...
+ */
+- (void)addressBarViewController:(ATLAddressBarViewController *)addressBarViewController retreivePrefilledParticipantsWithCompletion:(void (^)(NSArray *participants))completion;
```
Comment (kmahal) [inline Code/Controllers/ATLAddressBarViewController.h]: Rename to `setDefaultParticipantsWithCompletion:`

File: Code/Controllers/ATLAddressBarViewController.m (hunk 1 above)
Comment (dfinzer) [inline Code/Controllers/ATLAddressBarViewController.m]: Changed. Let me know if this looks better.

File: Code/Controllers/ATLAddressBarViewController.h (hunk 3 above)
Comment (dfinzer) [inline Code/Controllers/ATLAddressBarViewController.h]: Done.

File: Code/Controllers/ATLAddressBarViewController.h (hunk 2 above)
Comment (dfinzer) [inline Code/Controllers/ATLAddressBarViewController.h]: Done.

**Comment by dfinzer** [comment]
@kmahal thanks for taking a look. Addressed your comments.

File: Code/Controllers/ATLAddressBarViewController.m [hunk 4]
```diff
@@ -384,6 +382,27 @@ - (void)searchEnded
     self.participants = nil;
     self.tableView.hidden = YES;
     [self.tableView reloadData];
+    [self prefillParticipants];
+}
+
+- (void)prefillParticipants
```
Comment (kmahal) [inline Code/Controllers/ATLAddressBarViewController.m]: rename to `setDefaultParticipantsIfNecessary`

**Comment by kmahal** [comment]
@dfinzer Testing out the implementation - after selecting participants and sending a message to the conversation, the table view is not hidden, so the conversation can't be seen.  Please re-think the implementation for these use cases.  Additionally, please add a test in `ATLAddressBarControllerTest` for the new delegate method.

**Comment by dfinzer** [comment]
@kmahal fixed the issue you mentioned and added a test. Please take another look when you get a chance.

File: Code/Controllers/ATLAddressBarViewController.m [hunk 5]
```diff
@@ -384,6 +386,27 @@ - (void)searchEnded
...
     [self.tableView reloadData];
+    [self setDefaultParticipantsIfNecessary];
+}
+
+- (void)setDefaultParticipantsIfNecessary
+{
+    if ([self.delegate respondsToSelector:@selector(addressBarViewController:setDefaultParticipantsWithCompletion:)]) {
+        [self.delegate addressBarViewController:self setDefaultParticipantsWithCompletion:^(NSArray *participants) {
+            NSString *searchText = [self textForSearchFromTextView:self.addressBarView.addressBarTextView];
+            if (!searchText.length) {
```
Comment (kmahal) [inline Code/Controllers/ATLAddressBarViewController.m]: If the delegate takes time to respond with an array of objects and the user selects the input bar to send a message before the delegate responds, then the table view shows up on screen and won't go away.  Have a check here to see if the address bar is still first responder.

File: Code/Controllers/ATLAddressBarViewController.m [hunk 6]
```diff
@@ -253,15 +254,16 @@ - (void)textViewDidChange:(UITextView *)textView
...
-                self.participants = [self filteredParticipants:participants];
-                [self.tableView reloadData];
-                [self.tableView setContentOffset:CGPointZero animated:NO];
+                [self displayParticipants:participants];
             }];
         }
     }
 }
 
+- (void)textViewDidEndEditing:(UITextView *)textView {
```
Comment (kmahal) [inline Code/Controllers/ATLAddressBarViewController.m]: `{` on next line

**Comment by kmahal** [comment]
@dfinzer added a few more comments.

File: Code/Controllers/ATLAddressBarViewController.m (hunk 5 above)
Comment (dfinzer) [inline Code/Controllers/ATLAddressBarViewController.m]: Done.

File: Code/Controllers/ATLAddressBarViewController.m (hunk 6 above)
Comment (dfinzer) [inline Code/Controllers/ATLAddressBarViewController.m]: Done.

**Comment by dfinzer** [comment]
@kmahal addressed comments, thanks!

**Comment by dfinzer** [comment]
@kmahal any updates on whether this looks okay?

**Comment by dfinzer** [comment]
@kmahal @blakewatters any updates on whether this looks okay?

**Comment by kmahal** [comment]
@dfinzer I'll look at this today and provide any feedback, apologies for the delay.

**Comment by dfinzer** [comment]
@kmahal just want to check in on this pull request

**Comment by dfinzer** [comment]
@kmahal just want to check in on this, would love to get it merged in.

**Comment by dfinzer** [comment]
@kmahal any chance we could get this PR in? Been waiting for quite some time.

**Comment by kmahal** [comment]
@dfinzer after pulling this branch down it mostly looks good - however when the addressbar first becomes first responder the delegate is not called.  Only on subsequent typing and then deletion does it occur - once you get a fix for that we can merge it in.

**Comment by dfinzer** [comment]
@kmahal good catch. Fixed by placing the call in textViewDidBeginEditing

**Comment by kmahal** [comment]
@dfinzer Old branch but cleaning up Atlas right now.  One thing I noticed is that if the user scrolls on the pre-loaded page then the text view resigns first responder and the table view clears itself.  Get a fix in for that case and then we can bring this one in.  Feel free to close this PR if you are no longer working on this feature.

**Comment by kcoleman731** [comment]
@dfinzer  going to close this PR due to no response. Please let me know if you want to pick up the discussion,

**Pull request closed without merging**